### PR TITLE
Colorprofile

### PR DESCRIPTION
**Pull request by lrosenth**
@benjamingeer @BiancoA Modified the J2K-read method in order to process all but one of the conformance test files. In order to do so I added the ICC profiles into the include directory (which is allowed also by the Adobe license). Therefore the Adobe-profiles need no longer to be downloaded by cmake.

Conformance test file2.jp2 still poses problems which has to do with the processing of tiles within the standard kakadu method....

### CODE REVIEW

**Comment by benjamingeer** [comment]
The Travis build is failing:

```
Scanning dependencies of target icc_profiles
[ 69%] Generating ICC profile includes
cp: cannot stat '/usr/src/sipi/include/ICC-Profiles/AdobeRGB1998.icc': No such file or directory
make[2]: *** [../include/AdobeRGB1998_icc.h] Error 1
```

Is there something missing from the git repository?

**Comment by lrosenth** [comment]
Seems that .gitignore exclude files with *.icc extension. Changed this and uploaded the ICC profiles. Now waiting for the travis build result...

**Comment by lrosenth** [comment]
Ok, it compiles on travis ;-)

File: src/formats/SipiIOJ2k.cpp [hunk 1]
```diff
@@ -347,19 +349,54 @@ namespace Sipi {
...
-            int numcol = chaninfo.get_num_colours(); // I assume these are the color channels (1, 3 or 4 in case of CMYK)
-            for (size_t i = 0; i < img->nc - numcol; i++) { // img->nc - numcol: number of alpha channels (?)
-                img->es.push_back(ASSOCALPHA);
+            numcol = chaninfo.get_num_colours(); // I assume these are the color channels (1, 3 or 4 in case of CMYK)
+            int nluts = palette.get_num_luts();
+            if (nluts == 3) {
+                int nentries = palette.get_num_entries( );
+                rlut = new byte[nentries];
+                glut = new byte[nentries];
+                blut = new byte[nentries];
```
Comment (benjamingeer) [inline src/formats/SipiIOJ2k.cpp]: It looks like there's no `delete` to free the memory for these arrays.

File: src/formats/SipiIOJ2k.cpp [hunk 2]
```diff
@@ -443,6 +502,15 @@ namespace Sipi {
                 img->pixels = (byte *) buffer8;
                 break;
             }
+            case 12: {
+                bool *get_signed = new bool[img->nc];
```
Comment (benjamingeer) [inline src/formats/SipiIOJ2k.cpp]: It looks like the memory for `get_signed` is never freed.

File: src/formats/SipiIOJ2k.cpp [hunk 3]
```diff
@@ -443,6 +502,15 @@ namespace Sipi {
...
+                for (size_t i = 0; i < img->nc; i++) get_signed[i] = FALSE;
+                kdu_core::kdu_int16 *buffer16 = new kdu_core::kdu_int16[(int) dims.area() * img->nc];
+                decompressor.pull_stripe(buffer16, stripe_heights, nullptr, nullptr, nullptr, nullptr, get_signed);
+                img->pixels = (byte *) buffer16;
+                img->bps = 16;
+                break;
+            }
             case 16: {
                 bool *get_signed = new bool[img->nc];
```
Comment (benjamingeer) [inline src/formats/SipiIOJ2k.cpp]: It looks like the memory for `get_signed` is never freed.

**Comment by lrosenth** [comment]
fixed…

On 10 May 2017, at 11:21, Benjamin Geer <notifications@github.com<mailto:notifications@github.com>> wrote:


@benjamingeer requested changes on this pull request.

I can't test the conversions yet, but it looks like there are some memory leaks, could you just fix these? Then let's merge this and I hope to be able to test the conversions soon.

________________________________

In src/formats/SipiIOJ2k.cpp<https://github.com/dhlab-basel/Sipi/pull/157#discussion_r115691227>:

>          if (jpx_layer.exists()) {
             kdu_supp::jp2_colour colinfo = jpx_layer.access_colour(0);
             kdu_supp::jp2_channels chaninfo = jpx_layer.access_channels();
-            int numcol = chaninfo.get_num_colours(); // I assume these are the color channels (1, 3 or 4 in case of CMYK)
-            for (size_t i = 0; i < img->nc - numcol; i++) { // img->nc - numcol: number of alpha channels (?)
-                img->es.push_back(ASSOCALPHA);
+            numcol = chaninfo.get_num_colours(); // I assume these are the color channels (1, 3 or 4 in case of CMYK)
+            int nluts = palette.get_num_luts();
+            if (nluts == 3) {
+                int nentries = palette.get_num_entries( );
+                rlut = new byte[nentries];
+                glut = new byte[nentries];
+                blut = new byte[nentries];


It looks like there's no delete to free the memory for these arrays.

________________________________

In src/formats/SipiIOJ2k.cpp<https://github.com/dhlab-basel/Sipi/pull/157#discussion_r115691582>:

> @@ -443,6 +502,15 @@ namespace Sipi {
                 img->pixels = (byte *) buffer8;
                 break;
             }
+            case 12: {
+                bool *get_signed = new bool[img->nc];


It looks like the memory for get_signed is never freed.

________________________________

In src/formats/SipiIOJ2k.cpp<https://github.com/dhlab-basel/Sipi/pull/157#discussion_r115691672>:

> @@ -443,6 +502,15 @@ namespace Sipi {
                 img->pixels = (byte *) buffer8;
                 break;
             }
+            case 12: {
+                bool *get_signed = new bool[img->nc];
+                for (size_t i = 0; i < img->nc; i++) get_signed[i] = FALSE;
+                kdu_core::kdu_int16 *buffer16 = new kdu_core::kdu_int16[(int) dims.area() * img->nc];
+                decompressor.pull_stripe(buffer16, stripe_heights, nullptr, nullptr, nullptr, nullptr, get_signed);
+                img->pixels = (byte *) buffer16;
+                img->bps = 16;
+                break;
+            }
             case 16: {
                 bool *get_signed = new bool[img->nc];


It looks like the memory for get_signed is never freed.

**Pull request merged**